### DISCUSSION TITLE
Github will require ALL users to use 2FA???

### DISCUSSION THREAD

**Discussion by msharov** [community/community#16064] [Other Feature Feedback, Questions, & Ideas]
In your last blog entry you wrote:
```
GitHub will require all users who contribute code on
GitHub.com to enable one or more forms of two-factor
authentication (2FA) by the end of 2023.
```
Do you mean the users who contribute code *for* github.com *website*? Or will that be for all users contributing code to any repository hosted on github? Because if it is the latter, I'd appreciate a specific warning about it. 2FA is an absolutely "over my dead body" issue for me, and with this requirement I'd have to take all my repositories elsewhere. If I'll have to do it, the sooner I start the better.

**Comment by suprstarrd** [comment]
> 2FA is absolutely "over my dead body" issue for me, and with this requirement I'd have to take all my repositories elsewhere

I'm shocked that someone genuinely would dip from a service because of their attempts to increase security for every user. I'm not working with GitHub or anything, but this confuses me-- could you explain why you wouldn't want to use 2FA?

**Reply by myuanz** [reply]
I don't want to look at my phone when I log in, it's too troublesome. Now I don't care about the git repository in the account. If there is something to care about one day, I will naturally open 2FA by myself.

**Reply by suprstarrd** [reply]
@myuanz [Authy Desktop](https://authy.com/) may be of interest to you for the first point. It requires a phone number to set up, but from there all your 2FA can be on your desktop. 

As for you not caring about your git repositories, others have more sensitive repositories on their personal account (or an account under an organization that does not require 2FA) and at that point it is GitHub's responsibility to ensure that said repository is secure as it gets to a point of affecting a large open source ecosystem. I would say it's the user / organization, but if *GitHub* doesn't do anything, it makes *GitHub* look bad.

**Reply by myuanz** [reply]
Thanks for your reply, Authy Desktop is great. Before this I always thought 2FA could only be used on cell phones.

**Comment by msharov** [comment]
Because cell phones cost money. Upwards of $1000/year, which may be unaffordable to someone who does not have a silicon valley six figure salary. I'm shocked that you wouldn't know this. I have a smartphone, but only because T-Mobile still allows me to keep my $10/year legacy prepaid plan. If the plan is discontinued, I'll just have to do without the phone, and I wouldn't want to then lose access to all my published projects, would I?

Because 2FA is a huge PITA, turning a nearly instantaneous login process requiring only a single click on the login screen, into a five minute ordeal of getting the phone out of the desk, waiting two minutes for it to turn on, and then hurriedly entering three separate PINs (unlock phone, unlock Authy, type TOTP into GitHub). I'm shocked that anyone would think that this is simple and easy.

Because the security of my account is none of GitHub's business. Giving people an option to use 2FA is reasonable; requiring it is not. GitHub is not harmed in any way when my account is compromised, so what possible standing can they have to dictate to me how much work I have to do to authenticate? You say "it makes GitHub look bad", but that's not a reason; it's just an excuse for forcing people to do what GitHub wants. Dictatorships are indeed more secure, but why would I want to live in one?

Because 2FA does not improve security for everyone. Passwords get compromised mainly in three ways. First, when the account provider's site is hacked and the password hashes are cracked. Second, then this cracked password is reused elsewhere. And third, through phishing, when you type in your password where you shouldn't. I use strong passwords, unique for every site, and I store them in the password manager, which knows where it shouldn't type them. There is very little possiblity of my account being compromised already, so 2FA does not increase security for my account. It just increases the inconvenience by an order of magnitude.

Because there is a way to increase security without increasing the inconvenience. Use strong, unique passwords for every site. Stop thinking that passwords are something the user remembers and types in. If you think any normal person can remember a hundred strong passwords, you're insane. Making people think up more than two or three results in them reusing passwords, using simple weak ones, or just tweaking a single password many times. Just stop. Passwords should be stored in the password manager. The user should not know what the password is and should never be forced to type it. Remove this human factor, and you have very good security. No 2FA needed.

Of course, we are not always given a choice. If my bank or the government makes me do it, I don't have a choice. "Over my dead body" would just mean my dead body, and I don't feel suicidal right now. But with git hosting there are other options, and if GitHub insists on mandating any kind of phone-based 2FA, I'm taking my projects over to GitLab.

Then there is still a possibility of GitHub implementing a saner 2FA, such as FIDO U2F, which only requires touching my Yubikey during login as a minor inconvenience. Being tied to a $50 security key is at least better than to a $1000/year phone. If this is a possiblity I'll probably stay on GitHub. Mandating 2FA is stupid and mean, and I'll hate them for it, but the GitHub ecosystem is still better than GitLab's where there's nobody there. Of course, once GitHub starts mandating 2FA, many more people will go to GitLab, and this would become less of a drawback. We'll see.

**Reply by suprstarrd** [reply]
> Because cell phones cost money. Upwards of $1000/year, which may be unaffordable to someone who does not have a silicon valley six figure salary. I'm shocked that you wouldn't know this. I have a smartphone, but only because T-Mobile still allows me to keep my $10/year legacy prepaid plan. If the plan is discontinued, I'll just have to do without the phone, and I wouldn't want to then lose access to all my published projects, would I?

Valid point! This did slip my mind and is actually a good reason as to why this would be a bad idea. But for you specifically, this can be less of an issue (which I'll get to later).

> Because 2FA is a huge PITA, turning a nearly instantaneous login process requiring only a single click on the login screen, into a five minute ordeal of getting the phone out of the desk, waiting two minutes for it to turn on, and then hurriedly entering three separate PINs (unlock phone, unlock Authy, type TOTP into GitHub). I'm shocked that anyone would think that this is simple and easy.

As mentioned, Authy Desktop is still an option, which avoids the unlocking of the phone. And you don't need to do it quickly; because of how TOTP works, you can take as much time as you want, put in the PIN, and it will still be valid.

> Because the security of my account is none of GitHub's business. Giving people an option to use 2FA is reasonable; requiring it is not. GitHub is not harmed in any way when my account is compromised, so what possible standing can they have to dictate to me how much work I have to do to authenticate? You say "it makes GitHub look bad", but that's not a reason; it's just an excuse for forcing people to do what GitHub wants. Dictatorships are indeed more secure, but why would I want to live in one?

In my opinion, I do think it would be GitHub's business-- it's their job to keep users and customers secure and satisfied. But I know I can't convince anyone else otherwise (or vice versa), so this is still a valid point.

> Because 2FA does not improve security for everyone. Passwords get compromised mainly in three ways. First, when the account provider's site is hacked and the password hashes are cracked. Second, then this cracked password is reused elsewhere. And third, through phishing, when you type in your password where you shouldn't. I use strong passwords, unique for every site, and I store them in the password manager, which knows where it shouldn't type them. There is very little possiblity of my account being compromised already, so 2FA does not increase security for my account.

This is flat out incorrect-- it does. Regardless of "little possibility", there is still said possibility. GitHub could have a data breach at quite literally any point. Or what if someone shares a password with a friend? Or if you have a keylogger on your computer? These things can and do happen, and 2FA can pretty easily combat that (especially with what I'll mention later).

> Because there is a way to increase security without increasing the inconvenience. 

See above.

> Then there is still a possibility of GitHub implementing a saner 2FA, such as FIDO U2F, which only requires touching my Yubikey during login as a minor inconvenience. Being tied to a $50 security key is at least better than to a $1000/year phone. If this is a possiblity I'll probably stay on GitHub.

And here's where it's less of an issue: security keys are an option for GitHub 2FA (and arguably the most secure option!). I believe FIDO U2F should work but I haven't used a security key with my GitHub account so I have no way of verifying that.

**Reply by msharov** [reply]
> security keys are an option for GitHub 2FA

No, they are not. You are still required to do TOTP first. U2F can only be used as a third factor. There was a recent entry on the GitHub blog suggesting that they may implement U2F-only 2FA soon. I'm still waiting for news about that.

> In my opinion, I do think it would be GitHub's business--
> it's their job to keep users and customers secure and satisfied.

Sure, I would agree; but only through tasks relating to implementing their infrastructure properly. Mandating user behavior is well outside of that scope.

> Authy Desktop is still an option, which avoids the unlocking of the phone.

Authy still requires *having* a phone, which I pointed out was one of the big problems. Authy is also only available as a Snap package, which I am certainly not going to use, it is huge for a one-dialog application, and using it would still mean having to type in a PIN. Frequently typing PINs is still a major inconvenience, and it quickly comes back to the question of how much is being on GitHub worth to me? With Authy, I would say it is not. Everything I get here, I can get on GitLab, with the main exception being contributions from other developers. My projects don't get a lot of those anyway, so it really would not take all that much to push me over.

> Regardless of "little possibility", there is still said possibility. GitHub could 
> have a data breach at quite literally any point. Or what if someone shares a 
> password with a friend? Or if you have a keylogger on your computer? These things 
> can and do happen, and 2FA can pretty easily combat that

No, it can't. 2FA is not the perfect security that you have been brainwashed to believe. It is still vulnerable to phishing, and if it is not compromised as frequently, it is because there are so many people with weak passwords out there who make easier targets.

As for the other points, they are nonsense. I don't share my passwords, and can't, because I don't know what they are. My passwords are unique, so a data breach at GitHub would only breach my account there, and only if they foolishly store account passwords in plaintext, since the passwords are strong enough to not be practically crackable from hash. And as for a keylogger? Authy is the one vulnerable here, not me; I never type passwords - that's what the password manager is for.

Sure, there is still a possibility of compromise, just as there still is one with 2FA, but I am the one who should get to decide on the tradeoff between security and convenience, not GitHub.

**Reply by yan12125** [reply]
> Authy still requires having a phone

FWIW, I use KeePassXC for TOTP and it does not require a phone at all. I assume there are more tools like that. There are some concerns for such usage, though: https://github.com/keepassxreboot/keepassxc/issues/848

**Reply by RobMin** [reply]
Maan, you know there are other smart phones other than the latest Iphone? Also you can change your phone once every few years.
Look at this $50 for a smartphone and you can use it for several years
https://www.amazon.com/Tracfone-Motorola-Moto-Stylus-128GB/dp/B09CLNH78M/ref=sr_1_8?crid=14CRX6B57RMHT&keywords=smartphones&qid=1694591344&sprefix=smartphone%2Caps%2C307&sr=8-8

**Comment by ghost** [comment]
Unacceptable. I had to fill out 10 captchas only to get this error. Refresh the page and have to fill out 10 more. I basically said nope i'm not wasting time on this.
![rfgf](https://github.com/community/community/assets/31826503/f23ad3a1-aa28-4f32-9bd3-db1d947db370)
